dataplane/src/nat/mod.rs

```diff
@@ -0,0 +1,418 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+#![allow(dead_code)]
+
+mod fabric;
+mod iplist;
+mod prefixtrie;
+
+use crate::nat::fabric::{PeeringPolicy, Pif, Vpc};
+use crate::nat::iplist::{IpList, IpListType};
+use crate::nat::prefixtrie::PrefixTrie;
+use crate::packet::Packet;
+use crate::pipeline::NetworkFunction;
+
+use net::buffer::PacketBufferMut;
+use net::headers::Net;
+use net::headers::{TryHeadersMut, TryIpMut};
+use net::ipv4::UnicastIpv4Addr;
+use net::ipv6::UnicastIpv6Addr;
+use net::vxlan::Vni;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::net::IpAddr;
+
+#[derive(Debug)]
+#[allow(dead_code)]
+struct GlobalContext {
+    vpcs: HashMap<u32, Vpc>,
+    global_pif_trie: PrefixTrie,
+    peerings: HashMap<String, PeeringPolicy>,
+}
+
+impl GlobalContext {
+    #[tracing::instrument(level = "trace")]
+    fn new() -> Self {
+        Self {
+            vpcs: HashMap::new(),
+            global_pif_trie: PrefixTrie::new(),
+            peerings: HashMap::new(),
+        }
+    }
+
+    #[tracing::instrument(level = "trace")]
+    fn insert_vpc(&mut self, vni: Vni, vpc: Vpc) {
+        vpc.iter_pifs().for_each(|pif| {
+            pif.iter_ips().for_each(|prefix| {
+                let _ = self.global_pif_trie.insert(prefix, pif.name().clone());
```

Comment:
Shouldn't this check that a pif hasn't been already added for the prefix / ip?
Else, a pif may overwrite the ips of another one?

Reply:
Yes, absolutely. See my previous comment - I expect all of the code for object creation to move to some config component, where we'll make sure to enforce such safety checks. Right now these were mostly implemented for testing.